os/mac/clock_gettime.c

```diff
@@ -0,0 +1,56 @@
+#include <dlfcn.h>
+#include <inttypes.h>
+#include <pthread.h>
+#include <time.h>
+#include <sys/time.h>
+#if defined(__MACH__)
+#include <mach/mach_time.h>
+#else
+/* For compile testing on Linux */
+typedef struct mach_timebase_info {
+	uint32_t	numer;
+	uint32_t	denom;
+} mach_timebase_info_data_t;
+typedef struct mach_timebase_info *mach_timebase_info_t;
+uint64_t mach_absolute_time(void);
+int mach_timebase_info(mach_timebase_info_t info);
+#endif
+
+static typeof(clock_gettime) *clock_gettime_fn;
+static struct mach_timebase_info mach_timebase;
+static uint64_t mach_timestart;
+
+static void init_clock_gettime(void)
+{
+	void *libc = dlopen("/usr/lib/libc.dylib", RTLD_LAZY);
+	if (libc) {
+		clock_gettime_fn = dlsym(libc, "clock_gettime");
+		if (clock_gettime_fn)
+			return;
+	}
+	mach_timebase_info(&mach_timebase);
+	mach_timestart = mach_absolute_time();
+}
+
+int clock_gettime(clockid_t clk_id, struct timespec *tp)
+{
+	static pthread_once_t init = PTHREAD_ONCE_INIT;
+
+	pthread_once(&init, init_clock_gettime);
+	if (clock_gettime_fn)
+		return clock_gettime_fn(clk_id, tp);
+	if (clk_id == CLOCK_REALTIME) {
+		struct timeval tv;
+
+		if (gettimeofday(&tv, NULL) < 0)
+			return -1;
+		tp->tv_sec = tv.tv_sec;
+		tp->tv_nsec = tv.tv_usec * 1000;
+		return 0;
+	}
+	uint64_t delta_ns = (mach_absolute_time() - mach_timestart) *
+		(__uint128_t)mach_timebase.numer / mach_timebase.denom;
+	tp->tv_sec = delta_ns / (uint64_t)1e9;
+	tp->tv_nsec = delta_ns - tp->tv_sec * (uint64_t)1e9;
+	return 0;
+}
```

Comment:
My only thought is that modern macOS doesn't need this and really can do clock_gettime so we should only use this fallback when we don't have it. Would it be possible to guard this in a `#ifndef CONFIG_CLOCK_GETTIME`?

Reply:
> My only thought is that modern macOS doesn't need this and really can do clock_gettime so we should only use this fallback when we don't have it. Would it be possible to guard this in a `#ifndef CONFIG_CLOCK_GETTIME`?

This patch allows to remove the clock_gettime() configure check. Since the configure-time clock_gettime() check is fragile, I propose to remove that check instead of adding an #ifndef / #endif pair in this patch.

Reply:
With modern macOS supporting the regular code path we should support that as the first class citizen and deprioritise platform specific code. While it's not pretty I'd prefer one of the following:
- Use ifdefs around the above so modern macOS gets a regular path and older macOS gets a best effort path which can eventually be dropped
- Drop support for platforms without CLOCK_GETTIME. This would be done turning the check into a configure time failure, removing the CONFIG_CLOCK_GETTIME symbol and dropping code that is trying to work around the lack of it. [OS X 10.11 is five years old](https://en.wikipedia.org/wiki/OS_X_El_Capitan) and fio has already dropped support for Linux platforms that old. Further, macOS tools like homebrew have a policy of only supporting the last two macOS releases which means that [10.14 is the oldest macOS homebrew currently supports](https://github.com/Homebrew/brew/blob/1a9c3fa06e7fcc85239ec390ce1a0a4d268f3751/Library/Homebrew/os/mac.rb#L53) so we'd be in good company.